ceno_zkvm/src/instructions/riscv/mul.rs

```diff
@@ -213,15 +226,36 @@ impl<E: ExtensionField, I: RIVInstruction> Instruction<E> for MulhInstructionBas
                     rs2_read.value(),
                     rd_written.value(),
                     MulhSignDependencies::UU { constrain_rd },
+                    prod_low,
+                )
+            }
+            InsnKind::MUL => {
+                // constrain that prod_hi does not represent 2^32 - 1
+                let prod_hi_avoid = Expression::<E>::from(u32::MAX);
+                let prod_hi = UInt::new(|| "prod_hi", circuit_builder)?;
+                let constrain_rd = IsEqualConfig::construct_non_equal(
+                    circuit_builder,
+                    || "constrain_prod_hi",
+                    prod_hi.value(),
+                    prod_hi_avoid,
+                )?;
+
+                (
+                    rs1_read.value(),
+                    rs2_read.value(),
+                    rd_written.value(),
+                    MulhSignDependencies::LL { constrain_rd },
+                    prod_hi,
                 )
             }
 
             InsnKind::MULHSU => {
                 let rs1_signed = Signed::construct_circuit(circuit_builder, || "rs1", &rs1_read)?;
                 let rd_signed = Signed::construct_circuit(circuit_builder, || "rd", &rd_written)?;
+                let prod_low = UInt::new(|| "prod_low", circuit_builder)?;
 
                 // constrain that (signed) rd does not represent 2^31 - 1
-                let rd_avoid = Expression::<E>::from((1u64 << (BIT_WIDTH - 1)) - 1);
+                let rd_avoid = Expression::<E>::from(i32::MAX);
```

Comment:
Nice, I actually missed earlier that we are talking about `i32::MAX`, because the original arithmetic was a bit hard to read.  Thanks!